win: add UV_FS_O_FILEMAP

joaocgreis opened the pull request:
Writing files using a memory file mapping can be significantly faster on Windows. As an example, running npm with this option enabled can be up to 50% faster.

This is an opt-in because it can be a disadvantage in some situations, for example when doing multiple small write operations to the same file (in which case the buffering done by WriteFile is a big advantage). However, when writing files end-to-end, this can give a very noticeable performance improvement. In some situations (for example using external hard drives) the performance is the same, but not worst.

This PR doesn't currently have tests. Since this feature is opt-in, nothing should change and the current test suite should pass. If UV_FS_O_FILEMAP is enabled blindly for all writes, only 7 tests fail and for expected reasons. I'm still working on tests, will add them here when done or in a new PR if this one lands.

bzoz commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1386/

bnoordhuis commented on src/win/fs-fd-hash-inl.h:
```diff
@@ -0,0 +1,178 @@
+static struct uv__fd_hash_bucket_s uv__fd_hash[UV__FD_HASH_SIZE];
+
+
+INLINE static void uv__fd_hash_init() {
```
`uv__fd_hash_init(void)` - just `uv__fd_hash_init()` means a function taking any number of arguments.

bnoordhuis commented on src/win/fs-fd-hash-inl.h:
```diff
@@ -0,0 +1,178 @@
+    uv_fatal_error(err, "uv_mutex_init");
+  }
+
+  for (i = 0; i < UV__FD_HASH_SIZE; ++i) {
```
`i < ARRAY_SIZE(uv__fd_hash)`? General observation; rather than the constant try to use ARRAY_SIZE.

joaocgreis replied:
Done. Should I do the same for `UV__FD_HASH_GROUP_SIZE`?

bnoordhuis commented on src/win/fs-fd-hash-inl.h:
```diff
@@ -0,0 +1,178 @@
+#define FIND_IN_BUCKET_PTR()                                                 \
+  do {                                                                       \
+    size_t first_group_size = bucket_ptr->size % UV__FD_HASH_GROUP_SIZE;     \
+    if ((bucket_ptr->size != 0) && (first_group_size == 0))                  \
```
Superfluous parens. (Also on line 81, 125, 162, etc., by the way.)

bnoordhuis commented on src/win/fs-fd-hash-inl.h:
```diff
@@ -0,0 +1,178 @@
+
+  FIND_IN_BUCKET_PTR();
+
+  if (entry_ptr) {
```
Can you use explicit NULL checks here and on line 122, 128, etc.?

bnoordhuis commented on src/win/fs-fd-hash-inl.h:
```diff
@@ -0,0 +1,178 @@
+
+    if ((bucket_ptr->size != 0) && (i == 0)) {
+      struct uv__fd_hash_entry_group_s* new_group_ptr =
+        uv__malloc(sizeof(struct uv__fd_hash_entry_group_s));
```
`sizeof(*new_group_ptr)`

bnoordhuis commented on src/win/fs.c:
```diff
@@ -126,6 +127,8 @@
 #define IS_LETTER(c) (((c) >= L'a' && (c) <= L'z') || \
   ((c) >= L'A' && (c) <= L'Z'))
 
+#define MIN(a,b) (((a)<(b))?(a):(b))
```
Spaces around operators.

bnoordhuis commented on src/win/fs.c:
```diff
@@ -618,6 +704,117 @@ void fs__close(uv_fs_t* req) {
+
+
+void fs__read_filemap(uv_fs_t* req, struct uv__fd_info_s* fd_info) {
+  int fd = req->file.fd; // VERIFY_FD done in fs__read
```
C89 comments, please. Ditto line 898, 940.

bnoordhuis commented on src/win/fs.c:
```diff
@@ -618,6 +704,117 @@ void fs__close(uv_fs_t* req) {
+             (char*) view + view_offset + done_read,
+             this_read_size);
+    }
+    __except (fs__filemap_ex_filter(GetExceptionCode(),
```
When does this happen?

joaocgreis replied:
The exception is triggered when there is an error reading/writing the file. This follow the example from https://docs.microsoft.com/en-us/windows/desktop/Memory/reading-and-writing-from-a-file-view, but I used a filter to try to get some more information about the error.

joaocgreis commented:
@bnoordhuis thanks for your review, I updated the code. I added a new test for the hash table, will follow with more as soon as I get them cleaned up.

joaocgreis commented:
Tests added. Perhaps it would be good to run these tests on all platforms instead of only on Windows, but there are many differences that would be hard to track. For now, the tests show that the File Mapping version behaves exactly the same as the regular one.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1427/

@bnoordhuis if you could take another look, that would be great. Thanks!

joaocgreis commented:
Rebased and adapted to recent changes.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1454/
Node.js test build: https://nodejs.org/download/test/v12.6.0-test2019-07-107f1fedf9f0/

bzoz commented:
Landed in https://github.com/libuv/libuv/commit/2c279504f92a7ab1fc54c85ec36272722198a77e
